crates/bevy_pbr/src/render/mesh.rs

```diff
@@ -886,53 +898,73 @@ impl SpecializedMeshPipeline for MeshPipeline {
     }
 }
 
-#[derive(Resource)]
-pub struct MeshBindGroup {
-    pub normal: BindGroup,
-    pub skinned: Option<BindGroup>,
+/// Bind groups for meshes currently loaded.
+#[derive(Resource, Default)]
+pub struct MeshBindGroups {
+    model_only: Option<BindGroup>,
+    skinned: Option<BindGroup>,
+    morph_targets: HashMap<HandleId, BindGroup>,
+}
+impl MeshBindGroups {
+    pub fn reset(&mut self) {
+        self.model_only = None;
+        self.skinned = None;
+        self.morph_targets.clear();
+    }
+    /// Get the `BindGroup` for `GpuMesh` with given `handle_id`.
+    pub fn get(&self, handle_id: HandleId, is_skinned: bool, morph: bool) -> Option<&BindGroup> {
+        match (is_skinned, morph) {
+            (_, true) => self.morph_targets.get(&handle_id),
+            (true, false) => self.skinned.as_ref(),
+            (false, false) => self.model_only.as_ref(),
+        }
+    }
 }
 
 pub fn queue_mesh_bind_group(
-    mut commands: Commands,
+    meshes: Res<RenderAssets<Mesh>>,
+    mut groups: ResMut<MeshBindGroups>,
     mesh_pipeline: Res<MeshPipeline>,
     render_device: Res<RenderDevice>,
     mesh_uniforms: Res<ComponentUniforms<MeshUniform>>,
     skinned_mesh_uniform: Res<SkinnedMeshUniform>,
+    weights_uniform: Res<morph::Uniform>,
 ) {
-    if let Some(mesh_binding) = mesh_uniforms.uniforms().binding() {
-        let mut mesh_bind_group = MeshBindGroup {
-            normal: render_device.create_bind_group(&BindGroupDescriptor {
-                entries: &[BindGroupEntry {
-                    binding: 0,
-                    resource: mesh_binding.clone(),
-                }],
-                label: Some("mesh_bind_group"),
-                layout: &mesh_pipeline.mesh_layout,
-            }),
-            skinned: None,
-        };
-
-        if let Some(skinned_joints_buffer) = skinned_mesh_uniform.buffer.buffer() {
-            mesh_bind_group.skinned = Some(render_device.create_bind_group(&BindGroupDescriptor {
-                entries: &[
-                    BindGroupEntry {
-                        binding: 0,
-                        resource: mesh_binding,
-                    },
-                    BindGroupEntry {
-                        binding: 1,
-                        resource: BindingResource::Buffer(BufferBinding {
-                            buffer: skinned_joints_buffer,
-                            offset: 0,
-                            size: Some(NonZeroU64::new(JOINT_BUFFER_SIZE as u64).unwrap()),
-                        }),
-                    },
-                ],
-                label: Some("skinned_mesh_bind_group"),
-                layout: &mesh_pipeline.skinned_mesh_layout,
-            }));
+    groups.reset();
+
+    let layouts = &mesh_pipeline.mesh_layouts;
+
+    for (id, gpu_mesh) in meshes.iter() {
```

Comment:
This section is a bit overly complicated, largely because it moves some "global" work into the for loop. All logic hinges on the mesh uniforms buffer existing (aka if there is no mesh entity, we don't need to run any of this). `model_only` and `skinned` can both be checked for existence and computed outside of the for loop.

We only need to extract/check `Some(weights)` once outside of the for loop (as its only job should be to populate morph_targets (once we move out the other responsibilities). 

I'm preparing a commit.

Reply:
This works for me!